### PR TITLE
Remove /json from API URLs

### PR DESCRIPTION
**Pull request by bfirsh**
## What I did

Removed `/json` from API URLs.

I have left in the old URLs for backwards compatibility. They can be considered deprecated and removed at some point in the future.

[This is the corresponding engine-api change](https://github.com/docker/engine-api/pull/255), but there isn't a hard dependency on it because the URLs are backwards compatible.
## Why I did it
- They're all JSON anyway.
- If we want to add different formats, we can use the Accept header.
- Consistency with new `/networks` and `/volumes` endpoints.
- This was mentioned in https://github.com/docker/docker/issues/5893.
## Description for the changelog

Remove `/json` from all API URLs. For example, `GET /containers/json` is now just `GET /containers`.

## A picture of a cute animal (not mandatory but encouraged)

giphy-9

### CODE REVIEW

**Comment by thaJeztah** [comment]
:heart_eyes: sgtm!

**Comment by justincormack** [comment]
Distracted by the cats but sounds good, who wants ugly API endpoint names?

File: api/server/router/container/container.go [hunk 1]
```diff
@@ -41,15 +41,18 @@ func (r *containerRouter) initRoutes() {
 		// HEAD
 		router.NewHeadRoute("/containers/{name:.*}/archive", r.headContainersArchive),
 		// GET
-		router.NewGetRoute("/containers/json", r.getContainersJSON),
+		router.NewGetRoute("/containers", r.getContainers),
+		router.NewGetRoute("/containers/json", r.getContainers),
```
Comment (vdemeester) [inline api/server/router/container/container.go]: We should mark (and separate maybe ?) those with a comment to say they are deprecated and to not forgot to remove them at some point :angel:

File: docs/reference/api/docker_remote_api.md [hunk 2]
```diff
@@ -124,6 +123,11 @@ This section lists each version from latest to oldest.  Each listing includes a
 * `GET /images/json` now supports filters `since` and `before`.
 * `POST /containers/(id or name)/start` no longer accepts a `HostConfig`.
 * `POST /images/(name)/tag` no longer has a `force` query parameter.
+* `GET /containers/json` has been renamed to `/containers`.
```
Comment (vdemeester) [inline docs/reference/api/docker_remote_api.md]: We probably want to say the the old ones are deprecated and will not be available starting with version X :angel:

**Comment by vdemeester** [comment]
Design LGTM :+1:

**Comment by justincormack** [comment]
ok moving to code review then

**Comment by justincormack** [comment]
@bfirsh the test case failures look real to me...

**Comment by bfirsh** [comment]
@justincormack Yes, they are real. Sorry about that.

Another thing that occurred to me this morning is that this will stop containers and images called `json` from working. (Client will call `GET /containers/json` which will hit the deprecated route.)

I can't think of any option other than to completely remove those routes, which makes this much less of a no-brainer. :(

**Comment by thaJeztah** [comment]
@bfirsh we can remove them based on the API version that's being called

**Comment by thaJeztah** [comment]
err, guess that will need them to be forwarded to the "real" endpoint, using "json" as the requested name (if current version of the API is called)

**Comment by bfirsh** [comment]
Yeah, we can remove them based on API version, but that makes it a much less smooth upgrade path. Are there any objections to that?

**Comment by thaJeztah** [comment]
Hm, yes, thinking; we keep a 2 release deprecation policy usually, although for API it's versioned, so people can use the old API.

**Comment by cpuguy83** [comment]
IIRC, routes are hit in order that they are registered, may want to test this.
If so, you should be able to register the json endpoints first.

I don't think it's a good idea to remove the endpoints for the newer API... maybe if you push it out 2 versions but even still, it's breaking for not much of a reason.

**Comment by bfirsh** [comment]
@cpuguy83 These two routes will coexist if we maintain future compatibility: `/containers/{name:.*}` and `/containers/json`. I don't see how this is going to work if we also want to allow containers called `json`. ;)

I would argue API versioning allows us to make iterative breaking changes like this to the API, but I also realise it is a breaking change for what is arguably little reason.

My counter-argument is this is the first thing you see in the API documentation and has always made me think "gross" when looking at it. It makes users perceive Docker as poor quality, and is a [broken window](https://en.wikipedia.org/wiki/Broken_windows_theory) that encourages more bad design.

Then again, I care particularly about API design, so perhaps it doesn't matter in the general scheme of things. ;)

**Comment by justincormack** [comment]
I think this is not the most consistent approach, looking at the API again. The name `json` is ugly, I think it should be `inspect`, but I don't think the `inspect` output in the REST sense "is" the container, and there are a lot of issues with containers named `create` or `json`. I would add a new `inspect` API endpoint and deprecate `json`.

**Comment by LK4D4** [comment]
Removal of `json` will open possibility to switch gorilla/mux for lighter router :) Maybe not a big deal, but -4Mb of compiled regexps.

**Comment by justincormack** [comment]
There are other changes, `GET /images/search` would no longer work, conflict with images named `search`, it should probably be `GET /images?search=...` but I think this needs more planning and should skip to 1.13 alas, I think it is really high risk right now.

**Comment by thaJeztah** [comment]
I'm moving this back to design review while we discuss options

**Comment by thaJeztah** [comment]
ping @bfirsh did you have time to look into options?

**Comment by bfirsh** [comment]
@thaJeztah Yes, and so far it's looking much more complicated than I anticipated. ;)

Maybe we should close this and I'll open a separate proposal?

**Comment by thaJeztah** [comment]
Sounds good to me; perhaps just write up the new endpoints (old/new) and possible issues w.r.t. backward incompatibility.

Thanks!

**Pull request closed without merging**